kimchi/cairo/src/runner/bytecode.rs

```diff
@@ -0,0 +1,106 @@
+// TODO(querolita):
+// - be able to index memory with any type (u256 in particular) not only usize
+
+use crate::runner::word::CairoWord;
+
+/// This data structure stores the memory of the program
+pub struct CairoBytecode {
+    /// length of the public memory
+    publen: u64,
+    /// full memory vector, None if non initialized
+    pub vect: Vec<Option<CairoWord>>,
+}
+
+impl CairoBytecode {
+    /// Create a new memory structure from a vector of u64
+    pub fn new(input: Vec<i128>) -> CairoBytecode {
+        // Initialized with the public memory (compiled instructions only)
+        // starts intentionally with a zero word for ease of testing
+        let mut aux = vec![0];
+        aux.extend(input);
+        CairoBytecode {
+            publen: (aux.len() - 1) as u64,
+            vect: aux.into_iter().map(|i| Some(CairoWord::new(i))).collect(),
+        }
+    }
+
+    /// Get size of the public memory
+    pub fn public(&self) -> u64 {
+        self.publen
+    }
+
+    /// Get size of the full memory including dummy 0th entry
+    pub fn size(&self) -> u64 {
+        (self.vect.len()) as u64
+    }
+
+    /// Enlarges memory with enough additional None slots if necessary before writing or reading
+    fn enlarge(&mut self, index: u64) {
+        // if you want to access an index of the memory but its size is less or equal than this
+        if self.size() <= index {
+            // you will need to extend the vector with enough spaces (taking into account that
+            // vectors start by index 0 and size starts in 1)
+            let additional = index - self.size() + 1;
+            self.vect.reserve(additional.try_into().unwrap());
+            for _ in 0..additional {
+                // Consider CairoBytecode having Option<CairoWord> so one can have None here
+                self.vect.push(None);
+            }
+        }
+    }
+
+    /// Write u64 element in memory address
+    pub fn write(&mut self, index: u64, elem: i128) {
+        self.enlarge(index);
+        self.vect[index as usize] = Some(CairoWord::new(elem));
```

Comment:
maybe `index` should be renamed to `address` or `addr`?

Reply:
Oh you mean that it will panic if index is larger than usize? Meaning for instance, in a machine with 32-bit native words and casting a u64 into a usize directly? Or do you mean if the index is not yet allocated and one tries to access it? If it is the latter, the idea is that the `enlarge` (aka resize, as suggested in another comment) allocates enough memory to store enough cells so that accessing the index address shouldn't be a problem. And good suggestion for the `addr` name, I am changing it :)

Reply:
> Oh you mean that it will panic if index is larger than usize

no I meant if `index` is larger than the size of `vect`